### PR TITLE
Add Ant build support for XSpec for Schematron; fix Schematron compile bug

### PR DESCRIPTION
**Pull request by markdunnoup**
The build.xml file now runs XSpec for Schematron. The changes here also include some of the patches in the version of this file bundled with Oxygen XML Editor.

The "cleanup" target is not fully compatible with XSpec for Schematron yet. When ${clean.output.dir} is set to true, it removes SVRL files which the HTML report links to if the XSPec fails.

Also, I've fixed a bug that was preventing attribute rules from firing in Schematron. Not sure what the protocol for fixing this is. Presumably this code was cloned somehow from the Schematron repository.

### CODE REVIEW

**Comment by vincentml** [comment]
Thanks! I'll take a look at this in more detail as soon as I have some time. 

These changes to the Schematron XSLTs seem helpful. However, I don't think changes to the core Schematron XSLTs should be made here because that would make future updates to Schematron more difficult to maintain in XSpec. The entire src/schematron/iso_schematron folder is only imported. Would you consider sending a pull request to the [Schematron project ](https://github.com/Schematron/schematron) with your changes to the Schematron XSLTs?

**Comment by markdunnoup** [comment]
OK thanks. I was not sure of the protocol here. I've submitted the change as a pull request to the schematron/schematron repository.

**Comment by cirulls** [comment]
@markdunnoup, thanks for your contribution and thanks @vincentml  for reviewing it (I assigned this pull request to you). I agree on making changes related to Schematron in the Schematron project rather than here. Mark, does this pull request depend on the Schematron fix, i.e. do you need the Schematron issue to be fixed in order for the changed in the ant build to work? I guess the answer is no but it would be good if you can confirm.

Mark, if you post details on how you run the ant build (i.e. as [documented here](https://github.com/xspec/xspec/wiki/Running-with-Ant#ant-on-the-command-line) or do you need to pass any parameters?) and which version of ant you're currently using I can simulate the build and add a test in the XSpec test suite.

**Comment by markdunnoup** [comment]
I ran the Ant build in Oxygen 19.0 by creating a copy of Oxygen's default XSpec transformation scenario and amending it as follows:

- Under Options tab:
- change working directory to the folder containing the new build.xml file (e.g. C:\Git\xspec)
- Build file: [this file]
- Under Parameters tab:
- New parameter test.type (values "t" for testing XSLT and "s" for testing Schematron)
- Parameter xspec.project.dir value C:\Git\xspec
- New parameter xspec.compiled.xsl.dir (location to put XSLT file generated from Schematron file)
- Download xmltask.jar to C:\Git\xspec\bin\ (from http://www.oopsconsultancy.com/software/xmltask/)
- Add xmltask.jar to the Libraries list in the transformation scenario.

**Comment by cirulls** [comment]
Thanks Mark. I'll try to reproduce this on the command line so that we can add a test in the test suite for running ant.

File: build.xml [hunk 1]
```diff
@@ -38,89 +57,258 @@ where:
...
+  
+  <!-- Location of compiled Schematron XSL file 
+    Can be overridden, e.g. if original Schematron references helper files at relative paths, 
+    the compiled XSL file will need to go in the same folder as the original Schematron
+   -->
+  <property name="xspec.compiled.xsl.dir" value="${xspec.dir}"/>
+  <property name="xspec.compiled.xsl" value="${xspec.compiled.xsl.dir}/${xspec.base}-sch-compiled.xsl"/>
+  <!-- for passing as a parameter to the script that generates the XSpec with @stylesheet 
+  from the XSpec with @schematron -->
+  <property name="xspec.compiled.xsl.param" value="file:/${xspec.compiled.xsl}"/>
```
Comment (cirulls) [inline build.xml]: The slash in `file:/${xspec.compiled.xsl}` is problematic in Unix-like system (see full explanation in review). Try to change it with one of these two options and see which way (if any) works in Windows:
```
<property name="xspec.compiled.xsl.param" value="file:${xspec.compiled.xsl}"/>
<property name="xspec.compiled.xsl.param" value="file://${xspec.compiled.xsl}"/>
```

File: build.xml [hunk 2]
```diff
@@ -38,89 +57,258 @@ where:
...
+    <echo message="START COMPILING SCHEMATRON FILE '${xspec.schematron.absolute}'"/>
+    
+    <echo message="   STEP 1: Including modules"/>
+    
+    <xslt in="${xspec.schematron.absolute}" 
+      out="${xspec.dir}/${xspec.base}_dsdl_included.sch" 
+      style="${xspec.project.dir}/src/schematron/iso-schematron/iso_dsdl_include.xsl" 
+      force="true"
+      >
+      <factory name="net.sf.saxon.TransformerFactoryImpl"/>
```
Comment (cirulls) [inline build.xml]: Add the classpath instructions before this line:
```
<classpath>
    <pathelement path="${saxon.jar}"/>
    <pathelement path="${java.class.path}"/>
</classpath>
```

File: build.xml [hunk 3]
```diff
@@ -38,89 +57,258 @@ where:
...
+  <target name="compileSchematron" depends="locateSchematron" if="${test.schematron}">
+    <echo message="START COMPILING SCHEMATRON FILE '${xspec.schematron.absolute}'"/>
+    
+    <echo message="   STEP 1: Including modules"/>
+    
+    <xslt in="${xspec.schematron.absolute}" 
+      out="${xspec.dir}/${xspec.base}_dsdl_included.sch" 
+      style="${xspec.project.dir}/src/schematron/iso-schematron/iso_dsdl_include.xsl" 
+      force="true"
+      >
```
Comment (cirulls) [inline build.xml]: Add the classpath instructions here:
```
<classpath>
    <pathelement path="${saxon.jar}"/>
    <pathelement path="${java.class.path}"/>
</classpath>
```

File: build.xml [hunk 4]
```diff
@@ -38,89 +57,258 @@ where:
...
+      <factory name="net.sf.saxon.TransformerFactoryImpl"/>
+    </xslt>
+    
+    <echo message="   STEP 2: Expanding abstracts"/>
+    <xslt in="${xspec.dir}/${xspec.base}_dsdl_included.sch" 
+      out="${xspec.dir}/${xspec.base}.sch" 
+      style="${xspec.project.dir}/src/schematron/iso-schematron/iso_abstract_expand.xsl" 
+      force="true"
+      >
```
Comment (cirulls) [inline build.xml]: Add the classpath instructions here:
```
<classpath>
    <pathelement path="${saxon.jar}"/>
    <pathelement path="${java.class.path}"/>
</classpath>
```

File: build.xml [hunk 5]
```diff
@@ -38,89 +57,258 @@ where:
...
+      <factory name="net.sf.saxon.TransformerFactoryImpl"/>
+    </xslt>
+    
+    <echo message="   STEP 3: Convert Schematron to XSL"/>    
+    <xslt 
+      in="${xspec.dir}/${xspec.base}.sch"
+      out="${xspec.compiled.xsl}" 
+      style="${xspec.project.dir}/src/schematron/iso-schematron/iso_svrl_for_xslt2.xsl" 
+      force="true"
+      >
```
Comment (cirulls) [inline build.xml]: Add the classpath instructions here:
```
<classpath>
    <pathelement path="${saxon.jar}"/>
    <pathelement path="${java.class.path}"/>
</classpath>
```

File: build.xml [hunk 6]
```diff
@@ -38,89 +57,258 @@ where:
...
+      <param name="phase" expression="${xspec.phase}"/>
+    </xslt>
+    
+    <echo message="   STEP 4: Convert XSpec (into a format that references the XSL generated from the Schematron)"/>    
+    <xslt 
+      in="${xspec.xml}"
+      out="${xspec.compiled.xml}" 
+      style="${xspec.project.dir}/src/schematron/schut-to-xspec.xsl" 
+      force="true"
+      >
```
Comment (cirulls) [inline build.xml]: Add the classpath instructions here:
```
<classpath>
    <pathelement path="${saxon.jar}"/>
    <pathelement path="${java.class.path}"/>
</classpath>
```

File: build.xml [hunk 7]
```diff
@@ -8,26 +8,45 @@
...
+    
+    filename          XSpec tests XML file
+    
+    ---
+    
+    OUP patch for running XSpec 0.6.0 as an Oxygen XSLT transformation scenario
+    
+    This file is an updated copy of Oxygen's ${frameworksDir}/xspec/build.xml 
+    
+    Setup:
```
Comment (cirulls) [inline build.xml]: These instructions were very useful for the review. When the pull request is merged into `master` and documentation is added to the wiki I suggest to remove these comments and put a link to the wiki page instead.

File: build.xml [hunk 8]
```diff
@@ -38,89 +57,258 @@ where:
...
-				<param name="xspec.base.dir"
-							 location="${xspec.base.dir}" />
-				<param name="xspec.dir"
-							 location="${xspec.base.dir}/xspec" />
-				<param name="xspec.base"
-							 value="${xspec.base}" />
-			</antcall>
-		</sequential>
-	</macrodef>
+  <!-- OXYGEN PATCH needed to create a custom Saxon, for example  -warnings:recover -strip:none -opt:10 -dtd:off -l:off -versionmsg:off -expand:on -outval:fatal -val:lax -->
```
Comment (cirulls) [inline build.xml]: Once the code is merged into `master` I suggest to remove `OXYGEN PATCH` and `OUP PATCH` but keep all the relevant comments.

File: build.xml [hunk 9]
```diff
@@ -38,89 +57,258 @@ where:
...
+      
+    </xslt>
+    
+    <echo message="Running Tests..."/>
+    
+    <!-- OUP PATCH: Use Ant xslt task with Saxon initial template -->
+    <xslt in="${xspec.compiled.xml}" 
+      out="${xspec.result.xml}"
+      style="${xspec.xsl}"
+      force="true">
```
Comment (cirulls) [inline build.xml]: Add the classpath instructions here:
```
<classpath>
    <pathelement path="${saxon.jar}"/>
    <pathelement path="${java.class.path}"/>
</classpath>
```

File: src/schematron/iso-schematron/iso_schematron_skeleton_for_saxon.xsl [hunk 10]
```diff
@@ -516,7 +516,7 @@ which require a preprocess.
   which I find a bit surprising but anyway I'll use the longr faster version.
 -->
 <xsl:variable name="context-xpath">
-  <xsl:if test="$attributes='true' and parent::node() ">@*|</xsl:if>
+  <xsl:if test="$attributes='true'">@*|</xsl:if>
```
Comment (cirulls) [inline src/schematron/iso-schematron/iso_schematron_skeleton_for_saxon.xsl]: This concerns all the Schematron changes made in `iso_schematron_skeleton_for_saxon.xsl`: if they are not strictly needed to run the ant build file, I suggest to wait for these changes to be integrated in the [official Schematron repository](https://github.com/schematron/schematron), I saw you submitted a pull request there too.

**Comment by markdunnoup** [comment]
If the command line includes the saxon.jar file in the -lib parameter, the <classpath> instructions should not be required.

E.g. change

```
-Dsaxon.jar=/home/sandro/Documents/programming/saxon/saxon9he.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/xmltask-1.16.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/ant-contrib-1.0b3.jar 
```

to 

```
-lib /home/sandro/Documents/progra4mming/saxon/saxon9he.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/xmltask-1.16.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/ant-contrib-1.0b3.jar 
```

It might be possible to address the issue around the "file:" protocol in the XSLT, e.g.

```
    <xsl:variable name="filePrefix" as="xs:string*">
        <xsl:choose>
            <xsl:when test="contains(system-property('os.name'), 'Windows')">
                <xsl:text>file:/</xsl:text>
            </xsl:when>
            <xsl:otherwise>
                <xsl:text>file://</xsl:text>
            </xsl:otherwise>
        </xsl:choose>
    </xsl:variable>
```

and then concatenating this with the $stylesheet parameter:

```
<xsl:attribute name="stylesheet" select="concat($filePrefix,$stylesheet)"/>
```

In the build.xml file we would have just

```
<property name="xspec.compiled.xsl.param" value="${xspec.compiled.xsl}"/>
```

The proposed updates to Schematron itself are not a dependency for releasing this change.

**Comment by markdunnoup** [comment]
I've incorporated Sandro's comments. 

I did in the end have to use an XSLT $filePrefix variable to handle Windows/Linux variability.

**Comment by cirulls** [comment]
@markdunnoup: Sorry for the long wait on this. I finally had a chance to look at the failing tests in your latest commits and found out that the code that sets the variable `filePrefix` in `src/schematron/schut-to-xspec.xsl` is unnecessary. Removing that part allows to run ant for Schematron support correctly and makes the test suite pass for both Travis (Mac/Linux) and AppVeyor (Windows).

I took your latest changes done in fa109c9a103dafaa2ba0959d67c5d5200dc05744, removed the code related to `filePrefix`, and merged the `master` branch so that it picks up the latest changes made since you initiated the pull request. The code is in [branch review/138](https://github.com/xspec/xspec/tree/review/138) and the tests are now passing on [Travis](https://travis-ci.org/xspec/xspec/builds/332569305) and [AppVeyor](https://ci.appveyor.com/project/xspec/xspec/build/1.0.313). If you agree, I'll create a new pull request from branch `review/138` and merge it into `master` so that this feature can be used to run Schematron support via ant. This will be the last feature to be added before the next XSpec release (I'd like to release a new version before XML Prague).

I will add tests checking that the ant script works as intended and update the documentation in the wiki. For my own record, this is how I run ant from my machine (with hard-coded paths):

**In Linux**:

*XSLT support*:
```bash
ant 	-Dxspec.xml=tutorial/escape-for-regex.xspec \
	-lib /home/sandro/Documents/programming/saxon/saxon9he.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/xmltask-1.16.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/ant-contrib-1.0b3.jar
```

*Schematron support*:
```bash
ant 	-Dxspec.xml=tutorial/schematron/demo-03.xspec \
	-lib /home/sandro/Documents/programming/saxon/saxon9he.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/xmltask-1.16.jar \
	-lib /home/sandro/Documents/programming/github/xspec/bin/ant-contrib-1.0b3.jar \
	-Dtest.type=s \
	-Dxspec.compiled.xsl.dir=/home/sandro/Documents/programming/github/xspec/tutorial/schematron \
	-Dxspec.project.dir=/home/sandro/Documents/programming/github/xspec \
	-Dxspec.phase=#ALL
```

**In Windows**:

*XSLT support*:
```console
C:\apache-ant-1.9.9\bin\ant -Dxspec.xml=tutorial\escape-for-regex.xspec -Dsaxon.jar=C:\saxon\saxon9he.jar -lib C:\saxon\saxon9he.jar -lib C:\apache-ant-1.9.9\lib\xmltask-1.16.jar -lib C:\apache-ant-1.9.9\lib\ant-contrib-1.0b3.jar
```

*Schematron support*:
```console
C:\apache-ant-1.9.9\bin\ant -Dxspec.xml=tutorial\schematron\demo-03.xspec -Dsaxon.jar=C:\saxon\saxon9he.jar -lib C:\saxon\saxon9he.jar -lib C:\apache-ant-1.9.9\lib\xmltask-1.16.jar -lib C:\apache-ant-1.9.9\lib\ant-contrib-1.0b3.jar -Dtest.type=s -Dxspec.compiled.xsl.dir=C:\tmp\xspec\tutorial\schematron -Dxspec.project.dir=C:\tmp\xspec -Dxspec.phase=#ALL
```

**Comment by markdunnoup** [comment]
The change to remove the "filePrefix" code means that the Ant build won't run as a transformation scenario in Oxygen under Windows. (Oxygen doesn't recognize a file path beginning "C:/", only one beginning "file:/"

But maybe that just means that Oxygen will have to customize XSpec for Schematron for incorporation as a built-in scenario. That seems preferable to trying to fiddle with the continuous integration tools.

**Comment by cirulls** [comment]
I see, I completely forgot about the integration with oXygen! I suggest we go ahead with the merge of this feature for now, and I will discuss integration with oXygen with @AlexJitianu in February at XML Prague.

**Comment by cirulls** [comment]
@markdunnoup: this pull request has been included in #160 with modifications from me and @AirQuick and it is now merged into `master` branch. Thanks a lot for your precious work, this functionality is going to be very useful, especially for future integration with oXygen. Can I close this pull request since it is now included into `master`?

**Pull request closed without merging**